core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -345,13 +345,29 @@ private static RexNode canonizeNode(RexBuilder rexBuilder, RexNode condition) {
     case LESS_THAN_OR_EQUAL:
     case GREATER_THAN_OR_EQUAL: {
       RexCall call = (RexCall) condition;
-      final RexNode left = call.getOperands().get(0);
-      final RexNode right = call.getOperands().get(1);
+      RexNode left = call.getOperands().get(0);
+      left = canonizeNode(rexBuilder, left);
+      RexNode right = call.getOperands().get(1);
+      right = canonizeNode(rexBuilder, right);
+      call = (RexCall) rexBuilder.makeCall(call.getOperator(), left, right);
       if (left.toString().compareTo(right.toString()) <= 0) {
         return call;
       }
       return RexUtil.invert(rexBuilder, call);
     }
+    case PLUS:
+    case TIMES:
+      RexCall call = (RexCall) condition;
+      RexNode left = call.getOperands().get(0);
+      RexNode right = call.getOperands().get(1);
+      left = canonizeNode(rexBuilder, left);
+      right = canonizeNode(rexBuilder, right);
+
+      if (left.toString().compareTo(right.toString()) <= 0) {
+        return rexBuilder.makeCall(call.getOperator(), left, right);
+      } else {
+        return rexBuilder.makeCall(call.getOperator(), right, left);
```

Comment:
We should introduce an additional check before returning the new swapped expression to verify that the return type of the new call is similar to the return type of the original one (we can do that via `SqlTypeUtil.equalSansNullability`); otherwise, we should not swap. For instance, in Calcite the type inference for + or * will be dependent on the order of the operands. But other systems, e.g., Hive, may have their own implementation and apply different rules for type inference.